### PR TITLE
[W08H03] Use brute force to find possible edge cases

### PR DESCRIPTION
**Pull request by blubsbuar**
Vielleicht etwas viele überflüssige Tests, aber better be safe than sorry. Vergleicht die Rückgabe des Iterators mit einer von Collections.sort() sortierten ArrayList. Laut zulip entspricht das immer der korrekten Reihenfolge (https://zulip.in.tum.de/#narrow/stream/1499-PGdP-W08H03/topic/Baum.20Sortierung/near/852670).

### CODE REVIEW

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Random;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RandomTest {
+    @RepeatedTest(10000) // amount tests
+    public void randomlyTestIntValues() {
+        ArrayList<Integer> values = new ArrayList<>();
+        ArrayList<Integer> solution = new ArrayList<>();
+
+        Random num = new Random();
```
Comment (Anatoly03) [inline w08h03/test/pgdp/datastructures/RandomTest.java:19]: Bitte irgend ein Seed festlegen

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        assertEquals(actual, solution);
+    }
+
+    @RepeatedTest(10000) // amount tests
+    public void randomlyTestDoubleValues() {
+        ArrayList<Double> values = new ArrayList<>();
+        ArrayList<Double> solution = new ArrayList<>();
+
+        Random num = new Random();
```
Comment (Anatoly03) [inline w08h03/test/pgdp/datastructures/RandomTest.java:42]: hier auch

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @RepeatedTest(10000) // amount tests
+    public void randomlyTestStringValues() {
+        ArrayList<String> values = new ArrayList<>();
+        ArrayList<String> solution = new ArrayList<>();
+
+        int maxStringLength = 100; //max length of an individual string
+
+        Random num = new Random();
```
Comment (Anatoly03) [inline w08h03/test/pgdp/datastructures/RandomTest.java:90]: hier auch

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        assertEquals(actual, solution);
+    }
+
+    @RepeatedTest(10000) // amount tests
+    public void randomlyTestDateValues() {
+        ArrayList<Date> values = new ArrayList<>();
+        ArrayList<Date> solution = new ArrayList<>();
+
+        Random num = new Random();
```
Comment (Anatoly03) [inline w08h03/test/pgdp/datastructures/RandomTest.java:65]: hier auch

File: w08h03/test/pgdp/datastructures/RandomTest.java (hunk 1 above)
Comment (blubsbuar) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Warum? Der Test produziert keine Outputs die man vergleichen könnte und auch allgemein würde immer derselbe Seed doch dem wahllosen Testen von Bäumen schaden - dann würde einfach nur 10 000 mal derselbe Baum erstellt werden.

Comment (ipadmini22) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Sehe ich eigentlich wie @blubsbuar, aber es würde auch nicht wehtun.

Comment (nilsreichardt) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Wenn `Random` in Tests verwenden und den Seed nicht setzen, ist der Test nicht mehr deterministisch. Es könnte passieren, dass eine bestimmte Zahlenfolge nicht funktioniert, wodurch der Test failt. Man ändert an seiner Implementierung gar nichts, führt den Test wieder aus und auf einmal funktioniert der Test. Die Wahrscheinlichkeit bei 10.000 ist zwar gering, aber kann dennoch passieren.

Das Setzen eines Seeds macht den Test wieder deterministisch.

Comment (felixhauptmann) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: [Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)

Comment (blubsbuar) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: > Wenn wir `Random` in Tests verwenden und den Seed nicht setzen, ist der Test nicht mehr deterministisch. Es könnte passieren, dass eine bestimmte Zahlenfolge nicht funktioniert, wodurch der Test failt. Man ändert an seiner Implementierung gar nichts, führt den Test wieder aus und auf einmal funktioniert der Test. Die Wahrscheinlichkeit bei 10.000 ist zwar gering, aber kann dennoch passieren.
> 
> Das Setzen eines Seeds macht den Test wieder deterministisch.

Wenn der Test eine Zahlenfolge findet, für die dein Code nicht funktioniert, ist das doch im Grunde genommen genau das, worauf diese brute force Spielerei abzielt. Würde man hier einen fixen Seed festlegen, könnte man sich die 39996 übrigen Testausführungen sparen, weil immer dieselben Werte generiert werden würden. Die Tests berechnen die Lösung aber dynamisch (natürlich ohne, dass die Aufgabenstellung gelöst wird), weshalb sie eigentlich trotzdem noch korrekt sein müssen. Oder übersehe ich hier etwas?

Comment (blubsbuar) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: > [Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)

Auszug aus dem Link: "Fuzz testing, and therefore Autofixture, has a separate value: bug discovery. Here, randomization is actually desirable because it can help you find edge cases that you hadn't even anticipated." - Das ist doch genau, was ich mit diesen Tests versuche zu erreichen. Ein Seed würde einfach keinen Sinn machen.

Comment (nilsreichardt) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: >  Würde man hier einen fixen Seed festlegen, könnte man sich die 39996 übrigen Testausführungen sparen, weil immer dieselben Werte generiert werden würden.

Dafür ist ein Seed nicht da. 

Das feste Setzen eines Seed ist dafür da, dass man bestimmt, dass die Generierung von zufälligen Zahlen immer nach derselben Abfolge passiert. Sprich, wenn ich `Random(100).next(5)` aufrufe (Seed = 100), bekomme ich eine zufällige Zahl zwischen 0 und 4, z.B. 2. Dadurch ich denn Seed auf 100 festgelegt habe, weiß ich, dass ich wieder 2 bekomme, wenn ich `Random(100).next(5)` (ich erstelle ein _neues_ `Random` Objekt) aufrufe. Jedoch heißt das nicht, dass ich wieder 2 bekomme, ich mit _demselben_ `Random` Objekt `nextInt` aufrufe.

```java
import java.util.Random;

public class MyClass {
    public static void main(String args[]) {

      // Es ist wichtig, dass wir das Random Objekt nicht in der for-loop 
      // erstellen, sondern außerhalb
      Random r = new Random(100);

      for(int i = 0; i < 10; i++) {
          int randomInt = r.nextInt(1000);
          System.out.println(randomInt);
      }
    }
}
``` 

Du kannst den Code hier selber online ausprobieren: https://www.jdoodle.com/iembed/v0/BiO

Diese for-loop wird _immer_ als Zahlenfolge folgende Zahlen ausgeben: "915, 250, 874, 988, 291, 666, 36, 288, 723, 713". Jedoch ist jede Zahl an sich zufällig gewählt worden (es ist nicht im 915, 915, 915, 915), weil nur 1 `Ranom` Objekt erstellt haben. Würden wir *in* der for loop jedes mal ein neues `Random` Objekt erstellen, dann wäre es so wie du sagtest, dass jedes mal die Zahl 915 rauskommen würde.

Comment (blubsbuar) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Ja, das ist mir schon klar, aber würde ich einen Seed festlegen, dann würden meine Tests 10000 mal (wegen RepeatedTest(10000), was die Methode bzw. den Test 10000 mal aufruft -> also auch das Objekt 10000mal erstellt) 10000mal ein Random Objekt mit demselben Seed erstellen, wodurch alle 10000 Tests dieselben Zahlen testen  wie der 1.

Comment (nilsreichardt) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Wenn das Erstellen des `Random` Objektes aus dem Test ziehst, solltest du für jeden Test-Aufruf andere Wert bekommen, denn noch deterministisch die Werte erstellen können:

```java
Random randomTestIntValues = new Random();

@RepeatedTest(10000) // amount tests
public void randomlyTestIntValues() {
    ArrayList<Integer> values = new ArrayList<>();
    ArrayList<Integer> solution = new ArrayList<>();

    int amountOfValues = randomTestIntValues.nextInt(250); // max amount of values the tree should contain
    for (int i = 0; i < amountOfValues; i++) {
        values.add(randomTestIntValues.nextInt());
        solution = new ArrayList<>(values);
        Collections.sort(solution);
    }
    QuarternarySearchTree<Integer> n = new QuarternarySearchTree<Integer>();
    for (Integer i : values) {
        n.insert(i);
    }
    ArrayList<Integer> actual = new ArrayList<>();
    for (Integer r : n) {
        actual.add(r);
    }
    assertEquals(actual, solution);
}
```

Du müsstest nur darauf achten, dass du _nicht_ ein `Random` für einen weiteren Test (z.B. `randomlyTestDoubleValues()`) weiterverwendet, weil dies zu anderen Werten führen würde, wenn man die Tests in einer anderen Reihenfolge ausführt.

Comment (blubsbuar) [reply w08h03/test/pgdp/datastructures/RandomTest.java:19]: Gut, da hast du einen Punkt, werde es dahingehend anpassen

Comment (felixhauptmann) [reply w08h03/test/pgdp/datastructures/RandomTest.java]: > > [Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)
> 
> Auszug aus dem Link: "Fuzz testing, and therefore Autofixture, has a separate value: bug discovery. Here, randomization is actually desirable because it can help you find edge cases that you hadn't even anticipated." - Das ist doch genau, was ich mit diesen Tests versuche zu erreichen. Ein Seed würde einfach keinen Sinn machen.

Ich spreche mich ja auch nicht gegen randomisierte werte aus... Durch einen seed werden diese ja auch nicht weniger zufällig. Lediglich reproduzierbar zufällig, also das beste aus beiden Welten. Das Zitat stellt also kein Argument gegen einen festgelegten Seed dar.

**Review by nilsreichardt — COMMENTED** [review]
@blubsbuar Wie lange brauchen bei dir die 40.000 Tests zum durchlaufen? Bei mir braucht mein MacBook M1 Pro schon einige Minuten (ca. 5 Minuten). An sich die Tests definitiv sehr hilfreich (danke ❤️), aber ich glaube 40k sind doch etwas overkill :D

Ich denke, es reicht, wenn wir `@RepeatedTest(10000)` immer auf 1.000 setzen, also `@RepeatedTest(1000)`, womit wir insgesamt 4k tests haben. Das braucht ca. 10 Sekunden bei mir.

**Comment by ipadmini22** [comment]
@nilsreichardt also bei mir brauchend die 16 Sekunden

**Comment by nilsreichardt** [comment]
> Also bei mir brauchen die 16 Sekunden

Habs gerade mal mit IntelliJ probiert und da haben die 40.000 Tests auch nur 19 Sekunden gebraucht 🤔

Vielleicht können wir ein `final int numberOfTest = 1000;` unter den `Random` Variablen einfügen und in der Annotation dann `@RepeatedTest(numberOfTest)` verwenden. Dann könnte jeder mit der Änderung von einer Zeile die Anzahl der Tests ändern.

**Comment by DominikGleissner** [comment]
Ich bin gerade etwas verwirrt vom thread hier. wurden die tests schon gefixt bezüglich dem random seed? also sollten jetzt alle passen?

**Comment by nilsreichardt** [comment]
> Ich bin gerade etwas verwirrt vom thread hier. wurden die tests schon gefixt bezüglich dem random seed? also sollten jetzt alle passen?

Das mit dem Random ist schon gefixt, du kannst also einfach den Source Code von dem Test kopieren in dein Projekt einfügen :)

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.Random;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RandomTest {
+    Random intTest = new Random(100);
+    Random doubleTest = new Random(100);
+    Random dateTest = new Random(100);
+    Random stringTest = new Random(100);
+    
+    int numberOfTests = 1000;
```
Comment (nilsreichardt) [inline w08h03/test/pgdp/datastructures/RandomTest.java:19]:
```suggestion
    // If you want to increase the number of runs each test makes, just increase
    // this number.
    //
    // Please note that the test might be slow if you are using VS Code.
    // IntelliJ seems to be much faster.
    int numberOfTests = 1000;
```

Das wäre meine letzte Anmerkung, dass wir einen kurzen Kommentar schreiben und den Hinweis geben, dass VS Code etwas langsam sein könnte (so war zu mindestens bei mir gewesen), damit andere direkt Bescheid wissen.

Sobald wir das haben, können wir von meiner Seite aus mergen :)

**Review by nilsreichardt — CHANGES_REQUESTED** [review]

**Comment by DominikGleissner** [comment]
Und noch eine Anmerkung: Wenn ich die Tests so kopiere, sagt mir der Compiler, dass die Anzahl numberOfTests eine feste Zahl sein muss in z.B. @RepeatedTest(numberOfTests)... müsste numberOfTests nicht eigentlich final sein? Oder habe ich da was falsch?

**Comment by nilsreichardt** [comment]
> Und noch eine Anmerkung: Wenn ich die Tests so kopiere, sagt mir der Compiler, dass die Anzahl numberOfTests eine feste Zahl sein muss in z.B. @RepeatedTest(numberOfTests)... müsste numberOfTests nicht eigentlich final sein? Oder habe ich da was falsch?

Genau, es muss `final int numberOfTests = 1000;` sein

File: w08h03/test/pgdp/datastructures/RandomTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    Random doubleTest = new Random(100);
+    Random dateTest = new Random(100);
+    Random stringTest = new Random(100);
+    
+    final int numberOfTests = 1000;
+    // If you want to increase the number of runs per test, simply increase 
+    // this number.
+    //
+    // Please note that the tests might be slow, if you are using VS Code.
+    // IntelliJ seems to be much faster.
```
Comment (nilsreichardt) [inline w08h03/test/pgdp/datastructures/RandomTest.java:24]:
```suggestion
    // If you want to increase the number of runs per test, simply increase 
    // this number.
    //
    // Please note that the tests might be slow, if you are using VS Code.
    // IntelliJ seems to be much faster.
    final int numberOfTests = 1000;
```

Move comment above the variable.

**Review by nilsreichardt — APPROVED** [review]
LGTM

**Review by MaximilianAnzinger — COMMENTED** [review]
🐧
